code/modules/power/turbine/turbine.dm

```diff
@@ -0,0 +1,557 @@
+	var/produced_energy
+
+	///First stage of the moving gasmix - compression from 2500 L to 1000 L - heat up
+	var/datum/gas_mixture/compressor_mixture
+	///Second stage of the moving gasmix - expansion from 1000 L to 3000 L - first cool down
+	var/datum/gas_mixture/rotor_mixture
+	///Third stage of the moving gasmix - expansion from 3000 L to 6000 L - second cool down
+	var/datum/gas_mixture/turbine_mixture
+
+	///Check to see if all parts are connected to the core
+	var/all_parts_connected = FALSE
+	///If the machine was completed before reopening it, try to remake it
+	var/was_complete = FALSE
+
+	///Max rmp that the installed parts can handle, limits the rpms
+	var/max_allowed_rpm = 0
+	///Max temperature that the installed parts can handle, unlimited and causes damage to the machine
+	var/max_allowed_temperature = 0
+
+	///Amount of damage the machine has received
+	var/damage = 0
+	///Used to calculate the max damage received per tick and if the alarm should be called
+	var/damage_archived = 0
+
+	///Our internal radio
+	var/obj/item/radio/radio
+	///The key our internal radio uses
+	var/radio_key = /obj/item/encryptionkey/headset_eng
+	///The engineering channel
+	var/engineering_channel = "Engineering"
+
+	COOLDOWN_DECLARE(turbine_damage_alert)
+
+/obj/machinery/power/turbine/core_rotor/Initialize(mapload)
+	. = ..()
+	radio = new(src)
+	radio.keyslot = new radio_key
+	radio.set_listening(FALSE)
+	radio.recalculateChannels()
+
+/obj/machinery/power/turbine/core_rotor/LateInitialize()
+	. = ..()
+	activate_parts()
+
+/obj/machinery/power/turbine/core_rotor/Destroy(mob/user)
+	deactivate_parts(user)
+	QDEL_NULL(radio)
+	return ..()
+
+/obj/machinery/power/turbine/core_rotor/enable_parts(mob/user)
+	. = ..()
+	if(was_complete)
+		was_complete = FALSE
+		activate_parts(user)
+
+/obj/machinery/power/turbine/core_rotor/disable_parts(mob/user)
+	. = ..()
+	if(all_parts_connected)
+		was_complete = TRUE
+	deactivate_parts()
+
+/obj/machinery/power/turbine/core_rotor/multitool_act(mob/living/user, obj/item/tool)
+	if(!all_parts_connected && activate_parts(user))
+		balloon_alert(user, "all parts are linked")
+	return TOOL_ACT_TOOLTYPE_SUCCESS
+
+/obj/machinery/power/turbine/core_rotor/multitool_act_secondary(mob/living/user, obj/item/tool)
+	if(!all_parts_connected)
+		return TOOL_ACT_TOOLTYPE_SUCCESS
+	var/obj/item/multitool/multitool = tool
+	multitool.buffer = src
+	to_chat(user, span_notice("You store linkage information in [tool]'s buffer."))
+	return TOOL_ACT_TOOLTYPE_SUCCESS
+
+/obj/machinery/power/turbine/core_rotor/proc/activate_parts(mob/user)
+
+	compressor = locate(/obj/machinery/power/turbine/inlet_compressor) in get_step(src, turn(dir, 180))
+	turbine = locate(/obj/machinery/power/turbine/turbine_outlet) in get_step(src, dir)
+
+	var/parts_present = TRUE
+	if(!compressor || !turbine)
+		if(user)
+			balloon_alert(user, "missing parts detected")
+		parts_present = FALSE
+	if(compressor.dir != dir || !compressor.can_connect)
```

Comment:
The compressor/turbine existence check should not use parts_present. It should immediately return, since it will cause runtimes otherwise. See CI

Reply:
Right will fix it